### DISCUSSION TITLE
在`FormGrid`里如何通过外部状态设置`GridSpan`？

### DISCUSSION THREAD

**Discussion by Dogtiti** [alibaba/formily#3134] [Q&A 2.x]
demo：https://codesandbox.io/s/great-raman-9ke34p

直接通过`useState` 或者从 `props` 接收值无效，正确的写法应该是怎么写呢

![image](https://user-images.githubusercontent.com/38354472/169255567-7e4fab82-0880-41c5-a66a-29c1a3f22c13.png)

**Comment by janryWang** [comment]
https://formilyjs.org/zh-CN/guide/advanced/controlled

**Reply by Dogtiti** [reply]
不好意思例子有点不太准确，晚点我补一个，实际上的场景我是在做低代码的物料，那么`FormItem`是作为`Form`的`children`拖拽进去的，同时`setter`直接对`FormItem`,传递`props`,没有经过`Form`，(大概跟`formily designable`是一样的吧?)，这样的话相当于`FormItem`就是一个局部的片段，那么如何触发`Form`的更新呢，通常情况下所有的props都是从`Form`那一层传递下去，那么就可以`const form = React.useMemo(()=>createFrom(),[props])`，那么此时因为是`setter`直接透传`props`到`FormItem`,那么此时应该怎么做呢，或者最佳实践是什么呢

**Reply by Dogtiti** [reply]
```js
//正常情况下表单受控触发更新
const Form: React.FC<FormProps> = (props) => {
  const { children, ...restProps } = props;
  const { gridSpan} = restProps;

  const form = React.useMemo(() => createForm(), [gridSpan]);

  return (
    <Form form={form} colon={colon}>
      <FormGrid maxColumns={4} minColumns={4} columnGap={10}>
         <Field title={'title'} decorator={[FormItem, { gridSpan }]} name={'name'} component={[Input]} />
      </FormGrid >
    </Form>
  );
};
//低代码中由于<Field/>是物料，属性是通过`setter`面板直接透传过来的，没有经过`Form`,也就导致未受控，不会触发更新
const Form: React.FC<FormProps> = (props) => {
  const { children, ...restProps } = props
  const form = React.useMemo(() => createForm(), []);

  return (
    <Form form={form} colon={colon}>
       <FormGrid maxColumns={4} minColumns={4} columnGap={10}>
          {children}
       </FormGrid >
    </Form>
  );
};
//<FormInput/>作为物料直接被传递属性，导致无法更新
const FormInput: React.FC<ProFormInputProps> = (props) => {
  const { name = 'undefined', title = '表单项', gridSpan } = props;
  return (
    <Field title={title} decorator={[FormItem, { gridSpan }]} name={name} component={[Input]} />
  );
};
```
repo: https://github.com/Dogtiti/formily-sandbox.git
codesandbox: https://codesandbox.io/s/dogtiti-formily-sandbox-7o958z (好像不太好使，依赖比较慢)
执行`npm run lowcode:dev`启动，然后拖拽高级表单（已经配置好，最外层是`Form`,会有一个网格布局`FormGrid`,然后有一个`Input`）到画布.
点到网格布局配置,可以修改max,min columns
<img width="1429" alt="4625b99346a5168a99415bbc502c778" src="https://user-images.githubusercontent.com/38354472/169309716-46c3e244-e24c-40a9-b8a5-f515813fae92.png">
然后再点到输入框，可以配置列数
<img width="1429" alt="1fc2520905694ca6b708bebfc96baaa" src="https://user-images.githubusercontent.com/38354472/169310079-34062e3b-6054-4576-b479-6b9411219e1b.png">

此时`gridSpan`是不会触发更新的因为没受控，因为`gridSpan`是通过右侧`setter`直接传递到`FormItem`(输入框)，不经过Form`层，原因如上述，不知道这里处理的最佳实践是什么

`Form` 层：https://github.com/Dogtiti/formily-sandbox/blob/main/src/components/pro-form/pro-form.tsx
`FormGrid`层：https://github.com/Dogtiti/formily-sandbox/blob/main/src/components/pro-form-grid/pro-form-grid.tsx
`FormInput`层：https://github.com/Dogtiti/formily-sandbox/blob/main/src/components/pro-form-input/pro-form-input.tsx

**Reply by Dogtiti** [reply]
@janryWang 麻烦大佬有空帮忙看一下

**Comment by liuyisnake** [comment]
[https://codesandbox.io/s/restless-rain-stluje?file=/App.tsx](url)
